embedded-graphics/src/primitives/polyline.rs

```diff
@@ -0,0 +1,189 @@
+//! The line primitive
+
+use crate::draw_target::DrawTarget;
+use crate::drawable::Drawable;
+use crate::drawable::Pixel;
+use crate::geometry::Dimensions;
+use crate::geometry::Size;
+use crate::pixelcolor::PixelColor;
+use crate::primitives::Primitive;
+use crate::style::PrimitiveStyle;
+use crate::style::Styled;
+use crate::{
+    geometry::Point,
+    primitives::line::{Line, LineIterator},
+};
+
+/// Polyline primitive
+///
+/// Creates an unfilled chained line shape
+///
+/// # Examples
+///
+/// ## Create some lines with different styles
+///
+/// ```rust
+/// use embedded_graphics::{
+///     pixelcolor::Rgb565, prelude::*, primitives::Line, style::PrimitiveStyle,
+/// };
+/// # use embedded_graphics::mock_display::MockDisplay;
+/// # let mut display = MockDisplay::default();
+///
+/// // TODO: Example
+/// # Ok::<(), core::convert::Infallible>(())
+/// ```
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug, Default)]
+pub struct Polyline<'a> {
+    points: &'a [Point],
+}
+
+impl<'a> Polyline<'a> {
+    /// Create a new polyline from a list of points or an iterator
+    pub fn new(points: &'a [Point]) -> Self {
+        Self { points }
+    }
+
+    /// Get the total number of points in this polyline
+    pub fn len(&self) -> usize {
+        self.points.len()
+    }
+}
+
+impl<'a> Primitive for Polyline<'a> {}
+
+// TODO
+impl<'a> Dimensions for Polyline<'a> {
+    fn top_left(&self) -> Point {
+        Point::zero()
+    }
+
+    fn bottom_right(&self) -> Point {
+        self.top_left() + Point::zero()
+    }
+
+    fn size(&self) -> Size {
+        Size::from_bounding_box(self.top_left(), self.bottom_right())
+    }
+}
+
+/// TODO: Docs
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct PolylineIterator<'a> {
+    line: Polyline<'a>,
+    segment_iter: LineIterator,
+    start_idx: usize,
+    end_idx: usize,
+}
+
+impl<'a> Iterator for PolylineIterator<'a> {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
It might be possible to simplify this by using https://doc.rust-lang.org/std/primitive.slice.html#method.windows

Reply:
I tried this originally, but the code ends up something like this:

```rust
self.points
    .windows(2)
    .map(|[start, end, ..]| ThickLineIterator::new(&Line::new(*start, *end), 1))
    .flatten()
```

Storing that type in the iterator struct is _gnarly_, so I went with the hand made method instead. If you can figure a way to use `.windows()` instead that would be great.

Reply:
I think that stroing a closure inside a struct doesn't work at all. You would need to implement the mapping as a type:
```rust
struct LinesFromSlices<I> {
    iter: I,
}

impl<'a, I> Iterator for LinesFromSlices<I>
where
    I: Iterator<Item = &'a [Point]>,
{
    type Item = ThickLineIterator;

    fn next(&mut self) -> Option<Self::Item> {
        self.iter
            .next()
            .map(|[start, end, ..]| ThickLineIterator::new(&Line::new(*start, *end), 1))
    }
}
```
The type stored in `PolylineIterator` would be: `Flatten<LinesFromSlices<Windows<'a, Point>>>`

Reply:
That kind of works, however I have to do this:

```rust
#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
struct LinesFromSlices<I>(I);

impl<'a, I> Iterator for LinesFromSlices<I>
where
    I: Iterator<Item = &'a [Point]>,
{
    type Item = ThickLineIterator;

    fn next(&mut self) -> Option<Self::Item> {
        self.0.next().map(|parts| match parts {
            [start, end] => ThickLineIterator::new(&Line::new(*start, *end), 1),
            _ => unreachable!(),
        })
    }
}
```

I don't like having to add `unreachable!()` in there. Somehow `Windows` fixes that but I haven't looked into how/why.

We also lose nearly every derive except `Clone` and `Debug` when storing `Flatten<LinesFromSlices<Windows<'a, Point>>>` in a struct. Doesn't matter a great deal, but this is the only primitive with an iter that has this few derives IIRC.

IMO this change doesn't give much benefit over the "custom windows" implementation so I'd rather leave it as is. Is there a reason the existing code is dangerous or unclean that you'd like to see fixed?

Reply:
> IMO this change doesn't give much benefit over the "custom windows" implementation so I'd rather leave it as is. Is there a reason the existing code is dangerous or unclean that you'd like to see fixed?

No, I just thought it might result in easier to understand code.

Another way of simplifying would be by using a points slice in the iterator instead of the `Polyline`. The next line segment could always be constructed from the first two elements in the slice and then the first point could be removed from the slice. IMO that would be easier to understand than using the two index variables.